[DebugInfo] Fix line 0 records incorrectly having is_stmt set

RoshanYSingh23 opened the pull request:
Fixes issue #33870

Line 0 debug records should not have the is_stmt flag set in DWARF. This change ensures is_stmt is only set for non-zero line numbers.

Changes made in DwarfDebug.cpp:
- beginInstruction: Only set is_stmt for prologue_end if line != 0
- beginInstruction: Only set is_stmt for key instructions if line != 0
- emitInitialLocDirective: Set is_stmt=0 if scopeLine == 0

Test results: All 576 X86 DebugInfo tests pass.

OCHyams reviewed (commented):
Thanks for taking a look at the issue

Please could you remove unrelated formatting changes from the patch? It adds "noise" to reviews - typically we only format lines that are touched by the patch (git diff  -U0 | py -3 clang\tools\clang-format\clang-format-diff.py -p1 -binary "<path-to-llvm-build-dir>\bin\clang-format" -i).

> Test results: All 576 X86 DebugInfo tests pass.

As you're changing functionality, this patch will need to come with a/some new test/tests.

In the linked ticket it says

> There is one test related to prologue-end that seems to be testing that it's okay for prologue-end to be line 0, and I ran out of time to investigate any further.

Do you know which test that is? I wonder if there's any git-history which could be relevant

RoshanYSingh23 committed:
[DebugInfo] Fix line 0 records incorrectly having is_stmt set

Fixes issue #33870 (Bugzilla #34522)

Line 0 debug records should not have the is_stmt flag set in DWARF.
This change ensures is_stmt is only set for non-zero line numbers.

Changes made in DwarfDebug.cpp:
- beginInstruction: Only set is_stmt for prologue_end if line != 0
- beginInstruction: Only set is_stmt for key instructions if line != 0
- emitInitialLocDirective: Set is_stmt=0 if scopeLine == 0

Added two test cases to verify the behavior:
- llvm/test/DebugInfo/X86/line-0-no-is-stmt.ll: Verifies line 0 does not get is_stmt
- llvm/test/DebugInfo/X86/line-nonzero-has-is-stmt.ll: Verifies non-zero lines do get is_stmt

RoshanYSingh23 commented:
@OCHyams 
Thank you for the feedback! I've updated the patch to address all your concerns:

1. Removed formatting noise:
The patch now contains only the functional changes. Previously had 84 insertions/79 deletions, now streamlined to 74 insertions/4 deletions - all meaningful changes only.

2. Added test cases:
[llvm/test/DebugInfo/X86/line-0-no-is-stmt.ll]: Verifies that when scopeLine=0, the line 0 record does not have the is_stmt flag set
[llvm/test/DebugInfo/X86/line-nonzero-has-is-stmt.ll]: Verifies that when scopeLine=10, the record correctly has is_stmt and prologue_end flags set.
Both tests pass successfully.

3. Regarding the prologue-end test:
The existing prologue_end tests (e.g., [prologue_end.ll] all use non-zero scopeLine values (typically scopeLine: 2). These tests continue to work correctly.
The fix handles the edge case where prologue_end might occur at line 0 by conditionally setting is_stmt only when [DL.getLine() != 0]in the [beginInstruction] method. This ensures that if prologue_end does occur at line 0 (which is rare but possible), it won't incorrectly have is_stmt set.

I apologize for the initial sloppiness with the formatting - this is my first contribution to LLVM and I'm still learning the review process and best practices. If you have any other suggestions or concerns about the implementation or tests, I'd be very grateful to address them. Thank you for taking the time to review this!

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2199,12 +2199,16 @@ void DwarfDebug::beginInstruction(const MachineInstr *MI) {
 
   if (ScopeUsesKeyInstructions) {
-    if (IsKey)
+    // Don't set is_stmt for line 0
```
```suggestion
    // Don't set is_stmt for line 0.
```

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
As mentioned in a comment earlier, I feel like it's probably a bug if scope line is zero. But if we keep the other check for scope line in this patch, for completeness sake it should probably be replicated here.

OCHyams commented on llvm/test/DebugInfo/X86/line-0-no-is-stmt.ll:
```diff
@@ -0,0 +1,33 @@
+; CHECK: Address
+; CHECK: 0x{{[0-9a-f]+}} 0 0
+; CHECK-NOT: is_stmt
+; CHECK: end_sequence
```
Can we CHECK: the entire line instead? I feel that will be clearer to understand visually

OCHyams reviewed (commented):
Sorry it's taken a while to get back to you, thanks for making those changes. I've left a couple of nits and one question, but overall SGTM

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2400,8 +2404,11 @@ DwarfDebug::emitInitialLocDirective(const MachineFunction &MF, unsigned CUID) {
+  // However, we should not set is_stmt for line 0.
+  unsigned ScopeLine = SP->getScopeLine();
+  unsigned Flags = (ScopeLine != 0) ? DWARF2_FLAG_IS_STMT : 0;
+  ::recordSourceLine(*Asm, ScopeLine, 0, SP, Flags, CUID, getDwarfVersion(),
```
In which case I feel it's better to keep this code as it was previously, as it's almost certainly a bug if this is a line 0 and not particularly harmful if that occurs. So overall better to minimise change and complexity.

I don't feel massively strongly about it, as long as it's consistent between this line and the other scopeLine usage in this file, so if others feel differently I wouldn't insist.

OCHyams commented:
Reverse-ping, @RoshanYSingh23 are you in a position to finish this PR?
